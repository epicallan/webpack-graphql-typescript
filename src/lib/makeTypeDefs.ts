/**
 * looks for .gql or .graphql files
 * and builds out TS types for them
 * depends on @gql2ts/from-schema && merge-graphql-schemas
 * for core functionality
 */

import * as glob from 'glob-promise';
import { mergeTypes } from 'merge-graphql-schemas';
import { generateNamespace } from '@gql2ts/from-schema';
import * as fs from 'fs-extra';
import * as R from 'ramda';

export interface IgqlTsOpts {
    globPattern?: string;
    outFile?: string;
    nameSpace?: string;
}

const defaultOpts: IgqlTsOpts  = {
    globPattern: '**/*.gql',
    outFile: '',
    nameSpace: 'GQL'
};

const readFile: (fileName: string) => Promise<string> =
    (fileName) => fs.readFile(fileName, 'utf8');

const getGlobbedFilesLoad: (globPattern: string) => Promise<string[]> = async (globPattern) => {
    const files: string[]  = await glob(globPattern);
    const load: string[] = await Promise.all(R.map(readFile, files));
    return load;
};

export const getTypeDefs: (globPattern?: string) => Promise<string> = async (globPattern = '**/*.gql') => {
    try {
        const typesLoad: string[] = await getGlobbedFilesLoad(globPattern);
        const mergedTypes: string[] = mergeTypes(typesLoad);
        return R.join('', mergedTypes);
    } catch (error) {
        console.error(error);
    }
};

const addAutoGeneratedWarning = (tsTypes: string): string => {
    return `// This file is auto generated by the gqlToTs.ts module\n${tsTypes}`;
 };

export const generateTsFromGql: (options?: IgqlTsOpts) =>  Promise <string | any> = async (options = defaultOpts) => {
    try {
        const typeDefs: string = await getTypeDefs(options.globPattern);
        const namespaceOpts = { ignoreTypeNameDeclaration: true};
        const nameSpaceName = options.nameSpace || 'GQL';
        const tsNameSpace: string = generateNamespace(nameSpaceName, typeDefs, namespaceOpts, {});
        if (options.outFile && options.outFile.length) {
            fs.writeFile(options.outFile, addAutoGeneratedWarning(tsNameSpace));
        }
        return tsNameSpace;
    } catch (error) {
        console.error(error);
    }
};
