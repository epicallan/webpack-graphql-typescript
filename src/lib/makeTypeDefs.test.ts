import 'jest';
import { GraphQLSchema, buildSchema } from 'graphql';
import { generateTsFromGql, getTypeDefs} from './makeTypeDefs';
import * as prettyFormat from 'pretty-format';

describe('gql Types to Typescript types', () => {

    it('should return merged typedefs', async () => {
        const typeDefs = await getTypeDefs('**/*.gql');
        expect(prettyFormat(typeDefs)).toMatchSnapshot();
    });
    it('merged types should be convertable to a GraphQLSchema', async () => {
        const typeDefs = await getTypeDefs('**/*.gql');
        expect(buildSchema(typeDefs)).toBeInstanceOf(GraphQLSchema);
    });
    it.skip('should work with types with comments', async () => {
        // currently doesnt work with comments in the query type sections of schema
        const typeDefsWithComments =  await getTypeDefs('**/*.gql');
        expect(buildSchema(typeDefsWithComments)).toBeInstanceOf(GraphQLSchema);
        expect(prettyFormat(typeDefsWithComments)).toMatchSnapshot();
    });
    it ('End to End test: should create typescript types from graphql files', async () => {
        const tsTypesNameSpace = await generateTsFromGql({globPattern: '**/*.gql'});
        expect(prettyFormat(tsTypesNameSpace)).toMatchSnapshot();
    });
});
