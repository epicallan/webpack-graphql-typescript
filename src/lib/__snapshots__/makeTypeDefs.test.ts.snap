// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gql Types to Typescript types End to End test: should create typescript types from graphql files 1`] = `
"\\"// tslint:disable
// graphql typescript definitions

declare namespace GQL {
  interface IGraphQLResponseRoot {
    data?: IQuery | IMutation;
    errors?: Array<IGraphQLResponseError>;
  }

  interface IGraphQLResponseError {
    message: string;            // Required for all errors
    locations?: Array<IGraphQLResponseErrorLocation>;
    [propName: string]: any;    // 7.2.2 says 'GraphQL servers may provide additional entries to error'
  }

  interface IGraphQLResponseErrorLocation {
    line: number;
    column: number;
  }

  /*
    description: 
  */
  interface IQuery {
    getPerson: IPersonType | null;
    persons: Array<IPersonType> | null;
    getPersonWithExtra: IExtra | null;
  }

  /*
    description: 
  */
  interface IPersonType {
    name: string | null;
    id: string | null;
    sex: string | null;
  }

  /*
    description: 
  */
  interface IExtra {
    speed: number | null;
    person: IPersonType | null;
    agility: IAgilityMeasure | null;
  }

  /*
    description: 
  */
  interface IAgilityMeasure {
    value: number | null;
  }

  /*
    description: 
  */
  interface IMutation {
    addPerson: IPersonType | null;
  }
}

// tslint:enable
\\""
`;

exports[`gql Types to Typescript types should return merged typedefs 1`] = `
"\\"
    schema {
      query: Query
      mutation: Mutation
      
    }

    type Query {
     getPerson(id: String!): PersonType persons: [PersonType] getPersonWithExtra: Extra
  }

    type Mutation {
     addPerson(name: String!, sex: String!): PersonType
  }

    
  type PersonType {
  name: String
  id: String
  sex: String
}type AgilityMeasure {
  value: Int
}type Extra {
  speed: Int
  person: PersonType
  agility(height: Int): AgilityMeasure
}\\""
`;
